### PR TITLE
[AMD] Skip redundant CatArrayBatchedCopy in GLM-5 NSA TileLang decode

### PR DESCRIPTION
**Pull request by Jacob0226**
## Motivation

In GLM-5 NSA TileLang decode on ROCm, the fused-rope path dispatches a redundant `CatArrayBatchedCopy<OpaqueType<1u>, ...>` kernel once per layer per decode step that rebuilds an already-existing tensor.

The cause: `fused_qk_rope_cat_and_cache_mla` produces a contiguous `q_cat` of shape `(M, num_heads, kv_lora_rank + qk_rope_head_dim)`. The pre-patch flow then

1. slices `q_cat` into `q_nope_fused` / `q_pe_fused`,
2. passes them as separate args to `attn_mqa`,
3. `nsa_backend.forward_decode` calls `concat_mla_absorb_q_general(q_nope, q_rope)` to rebuild `q_all`,
4. on ROCm that helper falls back to `torch.cat([q_nope, q_rope], dim=-1)`, dispatching `CatArrayBatchedCopy` to produce a tensor byte-identical to the `q_cat` we already had.

The cat is pure overhead — same data, same layout, fresh allocation + copy.

## Modifications

(1) `forward_absorb_core` in `deepseek_common/attention_forward_methods/forward_mla.py` now passes `q_cat` directly to `attn_mqa` with `q_rope=None` on the decode path. Prefill keeps the split form because `nsa_backend.forward_extend` asserts `q_rope is not None`.

(2) `forward_decode` in `layers/attention/nsa_backend.py` now tracks `q_all` explicitly:
   - When the caller passes split `q_nope` / `q_rope`, `q_all = None` and each impl block re-cats as before — byte-identical to pre-patch behavior.
   - When the caller passes `q_rope=None` on HIP decode, `q_all` is set to a zero-copy `q.contiguous().view(...)` of `q_cat` and the cat is skipped.
   - The cat-skip is gated `if q_all is None or not _is_hip` so CUDA / MUSA / non-HIP backends always re-cat — those paths are bit-exact to pre-patch.

## Accuracy Tests

GLM-5.1-FP8 launch cmd:

```bash
export SAFETENSORS_FAST_GPU=1
export SGLANG_ROCM_FUSED_DECODE_MLA=0
export ROCM_QUICK_REDUCE_QUANTIZATION=INT4
python3 -m sglang.launch_server \
  --model GLM-5.1-FP8 \
  --tp 8 --port 8552 \
  --tool-call-parser glm47 --reasoning-parser glm45 \
  --watchdog-timeout 1200 \
  --mem-fraction-static 0.85 \
  --kv-cache-dtype fp8_e4m3 --disable-radix-cache \
  --model-loader-extra-config '{"enable_multithread_load": true, "num_threads": 8}' \
  --nsa-prefill-backend tilelang --nsa-decode-backend tilelang
```

MI355X GSM8K (TP=8, 1200 questions, parallel=1200): **0.941**

```bash
python3 /sgl-workspace/sglang/benchmark/gsm8k/bench_sglang.py \
    --port 8552 --num-questions 1200 --parallel 1200
```

## Speed Tests and Profiling

> **Baseline**: Thomas's PR #23562 (preshuffled paged MQA + `page_size=64`) applied on top of `sgl-project/sglang` main, plus aiter PR ROCm/aiter#2879. This PR is the delta on top of that baseline. The optimization is logically independent of #23562 — it only requires the `_use_aiter_gfx95 + nsa_decode_backend == "tilelang"` fast path which already exists on main — but #23562 is what we run as our perf-comparison reference.

Per-layer profiling (NSA TileLang decode on MI355X TP=8):

* `CatArrayBatchedCopy<OpaqueType<1u>, ...>` invocations per decode step: **390 -> 0** (one per layer eliminated).
* Saving: ~2.6 us / layer (constant; absolute, not relative).

Benchmark on MI355X TP=8, concurrency 4/8/16/32/64 averaged (`sglang.bench_serving --dataset-name random --random-range-ratio 0.8`):

* **Headline: Throughput +1.4% / TPOT -1.0% (faster)** vs Thomas's PR #23562 baseline, averaged across all 10 ISL/OSL/concurrency points.
* ISL/OSL 1k/1k: Throughput +0.96%, TPOT -1.31%
* ISL/OSL 8k/1k: Throughput +2.19%, TPOT -0.76%

Cat-skip's absolute ~2.6 us / layer benefit is constant; the relative gain is highest at small batch + short prompt (where total layer time is smallest) and decays with batch size.

Bench cmd template:

```bash
python3 -m sglang.bench_serving \
    --host localhost --port 8552 \
    --model /data/huggingface/hub/zai-org/GLM-5.1-FP8 \
    --dataset-name random \
    --random-input <8192|1024> --random-output 1024 --random-range-ratio 0.8 \
    --max-concurrency <4|8|16|32|64> --num-prompt <10*cc> \
    --output-file /dev/null
```

## Checklist

- [x] Format your code according to the [Format code](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code) with pre-commit (`pre-commit run --files <changed files>`: all hooks pass).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style guidance.

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the PR Merge Process.
2. Get approvals from CODEOWNERS and other reviewers.
3. Trigger CI tests with comments or contact authorized users to do so.
   * Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

### CODE REVIEW

File: python/sglang/srt/layers/attention/nsa_backend.py [hunk 1]
```diff
@@ -1573,8 +1579,12 @@ def forward_decode(
                 page_size=1,
             )
 
+        # Cat-skip is HIP-only: when caller passes q_rope=None on HIP, q_all has
+        # already been set to a zero-copy view of q in the else branch above and
+        # we can reuse it directly. Non-HIP backends fall through to the original
+        # "always cat" behavior to keep CUDA / MUSA paths byte-identical.
         if self.nsa_decode_impl == "flashmla_sparse":
-            if q_rope is not None:
+            if q_all is None or not _is_hip:
```
Comment (1am9trash) [inline python/sglang/srt/layers/attention/nsa_backend.py]: I think we can only make the change in amd side (e.g, tilelang/aiter backend). 
This change may be never reached in nv code path (e.g, flashmla_sparse/flashmla_kv).

Comment (Jacob0226) [reply python/sglang/srt/layers/attention/nsa_backend.py]: Good catch, thanks! Done in 1f2b7c48c.

Reverted the gate change for `flashmla_sparse` and `flashmla_kv` back to the pre-patch `if q_rope is not None:` pattern. Both impls import from `sgl_kernel.flashmla_ops` which requires CUDA driver, so on HIP they are unreachable and the new `not _is_hip` clause was a strict no-op there. On CUDA the behavior is byte-identical (gate is always True either way).

The cat-skip itself stays on the HIP-only `tilelang` and `aiter` blocks, which is what this PR is actually solving. Also moved the "Cat-skip is HIP-only ..." comment from above `flashmla_sparse` to above `tilelang` to match the new (narrower) scope.

**Review by 1am9trash — APPROVED** [review]
LGTM

**Review by kkHuang-amd — APPROVED** [review]
LGTM

**Comment by Jacob0226** [comment]
Quick CI triage: none of the failing jobs exercise this PR's code path.

The 6 unique failures are on `gpt-oss-20b`, `Qwen3-Reranker-0.6B`, `Llama-3.1-8B-Instruct`, `DeepSeek-OCR` (XPU), `LLaDA2-Mini` (NPU), and `Wan2.2 T2V` (NPU) — none of which use MLA + NSA TileLang decode on gfx95, which is the only path this PR touches. The cat-skip is also gated `if q_all is None or not _is_hip`, so CUDA / MUSA / XPU paths are byte-identical to pre-patch.

**Comment by HaiShaw** [comment]
@amd-bot ci-status
